// remix-auth-github/build/lib/emails.js
import { Request } from "./request.js";
export var Emails;
(function (Emails) {
    class Response {
        body;
        constructor(body) {
            this.body = body;
        }
        emails() {
            if (!Array.isArray(this.body))
                return [];
            let isValid = this.body.every((item) => {
                if (typeof item !== "object")
                    return false;
                if (item === null)
                    return false;
                if (Array.isArray(item))
                    return false;
                return ("email" in item &&
                    "verified" in item &&
                    "primary" in item &&
                    "visibility" in item);
            });
            if (!isValid)
                return [];
            let emails = this.body;
            return (emails
                .filter(({ verified }) => verified) // Filter out unverified emails
                // Sort to keep the primary email first
                .sort((a, b) => {
                if (a.primary && !b.primary)
                    return -1;
                if (!a.primary && b.primary)
                    return 1;
                return 0;
            })
                .map(({ email, primary }) => ({
                value: email,
                type: primary ? "primary" : "secondary",
            })));
        }
    }
    Emails.Response = Response;
    async function send(endpoint, context, options) {
        let request = context.toRequest(endpoint);
        let response = await fetch(request, { signal: options?.signal });
        let body = await response.json();
        let result = new Response(body);
        return result.emails();
    }
    Emails.send = send;
})(Emails || (Emails = {}));
