// remix-auth-oauth2/build/lib/authorization-code.js
/**
 * A lot of the code here was originally implemented by @pilcrowOnPaper for a
 * previous version of `@oslojs/oauth2`, as Pilcrow decided to change the
 * direction of the library to focus on response parsing, I decided to copy the
 * old code and adapt it to the new structure of the library.
 */
import { sha256 } from "@oslojs/crypto/sha2";
import { encodeBase64urlNoPadding } from "@oslojs/encoding";
export var AuthorizationCode;
(function (AuthorizationCode) {
    class AuthorizationURL extends URL {
        constructor(authorizationEndpoint, clientId) {
            super(authorizationEndpoint);
            this.searchParams.set("response_type", "code");
            this.searchParams.set("client_id", clientId);
        }
        setRedirectURI(redirectURI) {
            this.searchParams.set("redirect_uri", redirectURI);
        }
        addScopes(...scopes) {
            if (scopes.length < 1) {
                return;
            }
            let scopeValue = scopes.join(" ");
            const existingScopes = this.searchParams.get("scope");
            if (existingScopes !== null)
                scopeValue = ` ${existingScopes}`;
            this.searchParams.set("scope", scopeValue);
        }
        setState(state) {
            this.searchParams.set("state", state);
        }
        setS256CodeChallenge(codeVerifier) {
            const codeChallengeBytes = sha256(new TextEncoder().encode(codeVerifier));
            const codeChallenge = encodeBase64urlNoPadding(codeChallengeBytes);
            this.searchParams.set("code_challenge", codeChallenge);
            this.searchParams.set("code_challenge_method", "S256");
        }
        setPlainCodeChallenge(codeVerifier) {
            this.searchParams.set("code_challenge", codeVerifier);
            this.searchParams.set("code_challenge_method", "plain");
        }
    }
    AuthorizationCode.AuthorizationURL = AuthorizationURL;
})(AuthorizationCode || (AuthorizationCode = {}));
