// remix-auth-oauth2/build/lib/request.js
/**
 * A lot of the code here was originally implemented by @pilcrowOnPaper for a
 * previous version of `@oslojs/oauth2`, as Pilcrow decided to change the
 * direction of the library to focus on response parsing, I decided to copy the
 * old code and adapt it to the new structure of the library.
 */
import { encodeBase64 } from "@oslojs/encoding";
export var OAuth2Request;
(function (OAuth2Request) {
    class Context {
        method;
        body = new URLSearchParams();
        headers = new Headers();
        constructor(method) {
            this.method = method;
            this.headers.set("Content-Type", "application/x-www-form-urlencoded");
            this.headers.set("Accept", "application/json");
            this.headers.set("User-Agent", "oslo");
        }
        setClientId(clientId) {
            this.body.set("client_id", clientId);
        }
        authenticateWithRequestBody(clientId, clientSecret) {
            this.setClientId(clientId);
            this.body.set("client_secret", clientSecret);
        }
        authenticateWithHTTPBasicAuth(clientId, clientSecret) {
            const authorizationHeader = `Basic ${encodeBase64(new TextEncoder().encode(`${clientId}:${clientSecret}`))}`;
            this.headers.set("Authorization", authorizationHeader);
        }
        toRequest(url) {
            return new Request(url, {
                method: this.method,
                body: this.body,
                headers: this.headers,
            });
        }
    }
    OAuth2Request.Context = Context;
    // biome-ignore lint/suspicious/noShadowRestrictedNames: It's namespaced
    class Error extends globalThis.Error {
        request;
        context;
        description;
        uri;
        responseHeaders;
        constructor(message, request, context, responseHeaders, options) {
            super(message);
            this.request = request;
            this.context = context;
            this.responseHeaders = responseHeaders;
            this.description = options?.description ?? null;
            this.uri = options?.uri ?? null;
        }
    }
    OAuth2Request.Error = Error;
})(OAuth2Request || (OAuth2Request = {}));
