// remix-auth-oauth2/build/lib/token.js
/**
 * A lot of the code here was originally implemented by @pilcrowOnPaper for a
 * previous version of `@oslojs/oauth2`, as Pilcrow decided to change the
 * direction of the library to focus on response parsing, I decided to copy the
 * old code and adapt it to the new structure of the library.
 */
import { OAuth2RequestResult, TokenRequestResult } from "@oslojs/oauth2";
import { OAuth2Request } from "./request.js";
export var Token;
(function (Token) {
    let Request;
    (function (Request) {
        class Context extends OAuth2Request.Context {
            constructor(authorizationCode) {
                super("POST");
                this.body.set("grant_type", "authorization_code");
                this.body.set("code", authorizationCode);
            }
            setCodeVerifier(codeVerifier) {
                this.body.set("code_verifier", codeVerifier);
            }
            setRedirectURI(redirectURI) {
                this.body.set("redirect_uri", redirectURI);
            }
        }
        Request.Context = Context;
        async function send(endpoint, context, options) {
            let request = context.toRequest(endpoint);
            let response = await fetch(request, { signal: options?.signal });
            let body = await response.json();
            let result = new Result(body);
            if (result.hasErrorCode()) {
                throw new OAuth2Request.Error(result.errorCode(), request, context, response.headers, {
                    description: result.hasErrorDescription()
                        ? result.errorDescription()
                        : undefined,
                    uri: result.hasErrorURI() ? result.errorURI() : undefined,
                });
            }
            return result.toJSON();
        }
        Request.send = send;
        class Result extends TokenRequestResult {
            toJSON() {
                return {
                    ...this.body,
                    access_token: this.accessToken(),
                    token_type: this.tokenType(),
                    ...("expires_in" in this.body && {
                        expires_in: this.accessTokenExpiresInSeconds(),
                    }),
                    ...(this.hasScopes() && { scope: this.scopes().join(" ") }),
                    ...(this.hasRefreshToken() && { refresh_token: this.refreshToken() }),
                };
            }
        }
        Request.Result = Result;
    })(Request = Token.Request || (Token.Request = {}));
    let RevocationRequest;
    (function (RevocationRequest) {
        class Context extends OAuth2Request.Context {
            constructor(token) {
                super("POST");
                this.body.set("token", token);
            }
            setTokenTypeHint(tokenType) {
                if (tokenType === "access_token") {
                    this.body.set("token_type_hint", "access_token");
                }
                else if (tokenType === "refresh_token") {
                    this.body.set("token_type_hint", "refresh_token");
                }
            }
        }
        RevocationRequest.Context = Context;
        async function send(endpoint, context, options) {
            let request = context.toRequest(endpoint);
            let response = await fetch(request, { signal: options?.signal });
            let body = await response.json();
            let result = new OAuth2RequestResult(body);
            if (result.hasErrorCode()) {
                throw new OAuth2Request.Error(result.errorCode(), request, context, response.headers, {
                    description: result.hasErrorDescription()
                        ? result.errorDescription()
                        : undefined,
                    uri: result.hasErrorURI() ? result.errorURI() : undefined,
                });
            }
        }
        RevocationRequest.send = send;
    })(RevocationRequest = Token.RevocationRequest || (Token.RevocationRequest = {}));
    let RefreshRequest;
    (function (RefreshRequest) {
        class Context extends OAuth2Request.Context {
            constructor(refreshToken) {
                super("POST");
                this.body.set("grant_type", "refresh_token");
                this.body.set("refresh_token", refreshToken);
            }
            addScopes(...scopes) {
                if (scopes.length < 1) {
                    return;
                }
                let scopeValue = scopes.join(" ");
                const existingScopes = this.body.get("scope");
                if (existingScopes !== null) {
                    scopeValue = `${scopeValue} ${existingScopes}`;
                }
                this.body.set("scope", scopeValue);
            }
        }
        RefreshRequest.Context = Context;
    })(RefreshRequest = Token.RefreshRequest || (Token.RefreshRequest = {}));
})(Token || (Token = {}));
