// remix-auth-github/build/index.js
import { OAuth2Strategy, } from "remix-auth-oauth2";
import { Emails } from "./lib/emails.js";
import { Request } from "./lib/request.js";
import { UserProfile } from "./lib/user-profile.js";
export let GitHubStrategyDefaultName = "github";
export let GitHubStrategyDefaultScope = "user:email";
export let GitHubStrategyScopeSeperator = " ";
export class GitHubStrategy extends OAuth2Strategy {
    name = GitHubStrategyDefaultName;
    allowSignup;
    userAgent;
    userInfoEndpoint;
    userEmailsEndpoint;
    constructor({ allowSignup, userAgent, scopes = [GitHubStrategyDefaultScope], userInfoEndpoint = "https://api.github.com/user", userEmailsEndpoint = "https://api.github.com/user/emails", authorizationEndpoint = "https://github.com/login/oauth/authorize", tokenEndpoint = "https://github.com/login/oauth/access_token", ...options }, verify) {
        super({
            ...options,
            scopes,
            authorizationEndpoint,
            tokenEndpoint,
        }, verify);
        this.allowSignup = allowSignup ?? true;
        this.userAgent = userAgent ?? "Remix Auth";
        this.userInfoEndpoint = new URL(userInfoEndpoint);
        this.userEmailsEndpoint = new URL(userEmailsEndpoint);
    }
    authorizationParams(params) {
        let searchParams = new URLSearchParams(params);
        if (this.allowSignup)
            searchParams.set("allow_signup", "true");
        return searchParams;
    }
    async userEmails(accessToken) {
        let context = new Request.Context("GET", this.userAgent);
        context.authorize(accessToken);
        return await Emails.send(this.userEmailsEndpoint, context);
    }
    async userProfile(tokens) {
        let context = new Request.Context("GET", this.userAgent);
        context.authorize(tokens.access_token);
        let profile = await UserProfile.send(this.userInfoEndpoint, context);
        let emails = [{ value: profile.email }];
        if (this.options.scopes?.includes(GitHubStrategyDefaultScope)) {
            emails = await this.userEmails(tokens.access_token);
        }
        return {
            provider: "github",
            displayName: profile.login,
            id: profile.id.toString(),
            name: {
                familyName: profile.name,
                givenName: profile.name,
                middleName: profile.name,
            },
            emails: emails,
            photos: [{ value: profile.avatar_url }],
            _json: profile,
        };
    }
}
