// @oslojs/oauth2/dist/index.js
export class OAuth2RequestResult {
    body;
    constructor(body) {
        this.body = body;
    }
    hasErrorCode() {
        return "error" in this.body && typeof this.body.error === "string";
    }
    errorCode() {
        if ("error" in this.body && typeof this.body.error === "string") {
            return this.body.error;
        }
        throw new Error("Missing or invalid 'error' field");
    }
    hasErrorDescription() {
        return "error_description" in this.body && typeof this.body.error_description === "string";
    }
    errorDescription() {
        if ("error_description" in this.body && typeof this.body.error_description === "string") {
            return this.body.error_description;
        }
        throw new Error("Missing or invalid 'error_description' field");
    }
    hasErrorURI() {
        return "error_uri" in this.body && typeof this.body.error_uri === "string";
    }
    errorURI() {
        if ("error_uri" in this.body && typeof this.body.error_uri === "string") {
            return this.body.error_uri;
        }
        throw new Error("Missing or invalid 'error_uri' field");
    }
    hasState() {
        return "state" in this.body && typeof this.body.state === "string";
    }
    state() {
        if ("state" in this.body && typeof this.body.state === "string") {
            return this.body.state;
        }
        throw new Error("Missing or invalid 'state' field");
    }
}
export class TokenRequestResult extends OAuth2RequestResult {
    tokenType() {
        if ("token_type" in this.body && typeof this.body.token_type === "string") {
            return this.body.token_type;
        }
        throw new Error("Missing or invalid 'token_type' field");
    }
    accessToken() {
        if ("access_token" in this.body && typeof this.body.access_token === "string") {
            return this.body.access_token;
        }
        throw new Error("Missing or invalid 'access_token' field");
    }
    accessTokenExpiresInSeconds() {
        if ("expires_in" in this.body && typeof this.body.expires_in === "number") {
            return this.body.expires_in;
        }
        throw new Error("Missing or invalid 'expires_in' field");
    }
    accessTokenExpiresAt() {
        return new Date(Date.now() + this.accessTokenExpiresInSeconds() * 1000);
    }
    hasRefreshToken() {
        return "refresh_token" in this.body && typeof this.body.refresh_token === "string";
    }
    refreshToken() {
        if ("refresh_token" in this.body && typeof this.body.refresh_token === "string") {
            return this.body.refresh_token;
        }
        throw new Error("Missing or invalid 'refresh_token' field");
    }
    hasScopes() {
        return "scope" in this.body && typeof this.body.scope === "string";
    }
    scopes() {
        if ("scope" in this.body && typeof this.body.scope === "string") {
            return this.body.scope.split(" ");
        }
        throw new Error("Missing or invalid 'scope' field");
    }
}
export class DeviceAuthorizationRequestResult extends OAuth2RequestResult {
    deviceCode() {
        if ("device_code" in this.body && typeof this.body.device_code === "string") {
            return this.body.device_code;
        }
        throw new Error("Missing or invalid 'device_code' field");
    }
    userCode() {
        if ("user_code" in this.body && typeof this.body.user_code === "string") {
            return this.body.user_code;
        }
        throw new Error("Missing or invalid 'user_code' field");
    }
    verificationURI() {
        if ("verification_uri" in this.body && typeof this.body.verification_uri === "string") {
            return this.body.verification_uri;
        }
        throw new Error("Missing or invalid 'verification_uri' field");
    }
    codesExpireInSeconds() {
        if ("expires_in" in this.body && typeof this.body.expires_in === "number") {
            return this.body.expires_in;
        }
        throw new Error("Missing or invalid 'expires_in' field");
    }
    codesExpireAt() {
        return new Date(Date.now() + this.codesExpireInSeconds() * 1000);
    }
    intervalSeconds() {
        if ("interval" in this.body) {
            if (typeof this.body.interval === "number") {
                return this.body.interval;
            }
            throw new Error("Invalid 'interval' field");
        }
        return 5;
    }
}
